File: ios/SpellIt/Engine/BankStore.swift
import Foundation
import Observation

/// Custom word banks + the active-bank choice, persisted in UserDefaults.
@Observable
final class BankStore {
    static let shared = BankStore()

    private static let customKey = "spellit.customBanks"
    private static let activeKey = "spellit.activeBank"
    static let defaultBankId = "band-2-3"

    private struct StoredBank: Codable {
        var id: String
        var name: String
        var entries: [WordEntry]
    }

    private(set) var customBanks: [WordBank] = []
    private(set) var activeId: String = BankStore.defaultBankId

    private init() {
        if let data = UserDefaults.standard.data(forKey: Self.customKey),
           let stored = try? JSONDecoder().decode([StoredBank].self, from: data) {
            customBanks = stored.map {
                WordBank(
                    id: $0.id,
                    name: $0.name,
                    blurb: "Your custom list",
                    builtIn: false,
                    entries: $0.entries,
                )
            }
        }
        if let saved = UserDefaults.standard.string(forKey: Self.activeKey),
           allBanks.contains(where: { $0.id == saved }) {
            activeId = saved
        }
    }

    var allBanks: [WordBank] {
        WordData.builtInBanks + customBanks
    }

    var activeBank: WordBank {
        allBanks.first { $0.id == activeId }
            ?? WordData.builtInBanks.first { $0.id == Self.defaultBankId }
            ?? WordData.builtInBanks[0]
    }

    func setActive(_ id: String) {
        activeId = id
        UserDefaults.standard.set(id, forKey: Self.activeKey)
    }

    private func persist() {
        let stored = customBanks.map {
            StoredBank(id: $0.id, name: $0.name, entries: $0.entries)
        }
        if let data = try? JSONEncoder().encode(stored) {
            UserDefaults.standard.set(data, forKey: Self.customKey)
        }
    }

    @discardableResult
    func createBank(name: String = "My new list", entries: [WordEntry] = []) -> WordBank {
        let bank = WordBank(
            id: UUID().uuidString,
            name: name,
            blurb: "Your custom list",
            builtIn: false,
            entries: entries,
        )
        customBanks.append(bank)
        persist()
        return bank
    }

    func deleteBank(_ id: String) {
        customBanks.removeAll { $0.id == id }
        if activeId == id { setActive(Self.defaultBankId) }
        persist()
    }

    func rename(_ id: String, to name: String) {
        guard let index = customBanks.firstIndex(where: { $0.id == id }) else { return }
        customBanks[index].name = name
        persist()
    }

    func addWord(_ entry: WordEntry, to id: String) {
        guard let index = customBanks.firstIndex(where: { $0.id == id }) else { return }
        customBanks[index].entries.append(entry)
        persist()
    }

    func removeWord(at wordIndex: Int, from id: String) {
        guard let index = customBanks.firstIndex(where: { $0.id == id }),
              customBanks[index].entries.indices.contains(wordIndex)
        else { return }
        customBanks[index].entries.remove(at: wordIndex)
        persist()
    }
}

// MARK: - Word-length heuristics (banks carry no grade)

enum GameHeuristics {
    /** How many letters Missing Letters blanks out. */
    static func blanks(for word: String) -> Int {
        min(4, max(1, Int((Double(word.count) / 3).rounded())))
    }

    /** How long Flash Spell shows the word before hiding it. */
    static func flashSeconds(for word: String) -> Double {
        if word.count <= 4 { return 4 }
        if word.count <= 7 { return 3.5 }
        return 3
    }
}

File: android/app/src/main/java/com/skdaddle/spellit/engine/BankStore.kt
package com.skdaddle.spellit.engine

import android.content.Context
import android.content.SharedPreferences
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.skdaddle.spellit.data.WordData
import com.skdaddle.spellit.model.WordBank
import com.skdaddle.spellit.model.WordEntry
import java.util.UUID
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/** Custom word banks + the active-bank choice, persisted in SharedPreferences. */
class BankStore private constructor(private val prefs: SharedPreferences) {
    companion object {
        private const val PREFS_NAME = "spellit"
        private const val CUSTOM_KEY = "spellit.customBanks"
        private const val ACTIVE_KEY = "spellit.activeBank"
        const val DEFAULT_BANK_ID = "band-2-3"

        @Volatile
        private var instance: BankStore? = null

        fun shared(context: Context): BankStore =
            instance ?: synchronized(this) {
                instance ?: BankStore(
                    context.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE),
                ).also { instance = it }
            }
    }

    @Serializable
    private data class StoredBank(
        val id: String,
        val name: String,
        val entries: List<WordEntry>,
    )

    private val json = Json { ignoreUnknownKeys = true }

    var customBanks by mutableStateOf<List<WordBank>>(emptyList())
        private set
    var activeId by mutableStateOf(DEFAULT_BANK_ID)
        private set

    /** Bumps on any custom-bank mutation so mid-round edits re-deal games. */
    var revision by mutableIntStateOf(0)
        private set

    init {
        val data = prefs.getString(CUSTOM_KEY, null)
        if (data != null) {
            val stored = runCatching { json.decodeFromString<List<StoredBank>>(data) }
                .getOrDefault(emptyList())
            customBanks = stored.map {
                WordBank(
                    id = it.id,
                    name = it.name,
                    blurb = "Your custom list",
                    builtIn = false,
                    entries = it.entries,
                )
            }
        }
        val saved = prefs.getString(ACTIVE_KEY, null)
        if (saved != null && allBanks.any { it.id == saved }) {
            activeId = saved
        }
    }

    val allBanks: List<WordBank>
        get() = WordData.builtInBanks + customBanks

    val activeBank: WordBank
        get() = allBanks.firstOrNull { it.id == activeId }
            ?: WordData.builtInBanks.firstOrNull { it.id == DEFAULT_BANK_ID }
            ?: WordData.builtInBanks[0]

    fun setActive(id: String) {
        activeId = id
        prefs.edit().putString(ACTIVE_KEY, id).apply()
    }

    private fun persist() {
        revision += 1
        val stored = customBanks.map { StoredBank(it.id, it.name, it.entries) }
        prefs.edit().putString(CUSTOM_KEY, json.encodeToString(stored)).apply()
    }

    fun createBank(name: String = "My new list", entries: List<WordEntry> = emptyList()): WordBank {
        val bank = WordBank(
            id = UUID.randomUUID().toString(),
            name = name,
            blurb = "Your custom list",
            builtIn = false,
            entries = entries,
        )
        customBanks = customBanks + bank
        persist()
        return bank
    }

    fun deleteBank(id: String) {
        customBanks = customBanks.filterNot { it.id == id }
        if (activeId == id) setActive(DEFAULT_BANK_ID)
        persist()
    }

    fun rename(id: String, name: String) {
        val index = customBanks.indexOfFirst { it.id == id }
        if (index < 0) return
        customBanks = customBanks.toMutableList().also {
            it[index] = it[index].copy(name = name)
        }
        persist()
    }

    fun addWord(entry: WordEntry, bankId: String) {
        val index = customBanks.indexOfFirst { it.id == bankId }
        if (index < 0) return
        customBanks = customBanks.toMutableList().also {
            it[index] = it[index].copy(entries = it[index].entries + entry)
        }
        persist()
    }

    fun removeWord(wordIndex: Int, bankId: String) {
        val index = customBanks.indexOfFirst { it.id == bankId }
        if (index < 0) return
        val entries = customBanks[index].entries
        if (wordIndex !in entries.indices) return
        customBanks = customBanks.toMutableList().also {
            it[index] = it[index].copy(
                entries = entries.toMutableList().apply { removeAt(wordIndex) },
            )
        }
        persist()
    }
}

// Word-length heuristics (banks carry no grade)

object GameHeuristics {
    /** How many letters Missing Letters blanks out. */
    fun blanks(word: String): Int =
        minOf(4, maxOf(1, Math.round(word.length / 3.0).toInt()))

    /** How long Flash Spell shows the word before hiding it, in milliseconds. */
    fun flashMs(word: String): Long = when {
        word.length <= 4 -> 4000L
        word.length <= 7 -> 3500L
        else -> 3000L
    }
}
